File: backend/app/export.py
"""Markdown report rendered server-side from a run view."""
from datetime import datetime, timezone

from .config import region_name


def export_markdown(view: dict) -> str:
    run = view["run"]
    evidence = view["evidence"]
    day = datetime.fromtimestamp((run.get("finishedAt") or run["createdAt"]) / 1000, tz=timezone.utc).strftime("%Y-%m-%d")
    out = ["# LaunchRadar — %s" % run["question"], "",
           "- Region: %s" % region_name(run["region"]), "- Run: %s · %s" % (run["id"], day),
           "- Searches used: %d / budget %d" % (run["searchesUsed"], run["budget"]),
           "- Evidence: %d rows, %d domains, %d source types" % (
               len(evidence), len({e["domain"] for e in evidence if e.get("domain")}), len({e["blockType"] for e in evidence})), "",
           "## Opportunities (ranked)", ""]
    opps = sorted(view["opportunities"], key=lambda o: o["score"], reverse=True)
    if not opps:
        out += ["_None — every candidate gap is already served._", ""]
    for i, o in enumerate(opps):
        s = o["subScores"]
        out += ["### %d. %s" % (i + 1, o["title"]), "",
                "Score **%s/100** (%s). Pain %s, Momentum %s, Commercial %s, Whitespace %s, Weak rivals %s." % (
                    o["score"], o["confidence"], s["pain"], s["momentum"], s["commercial"], s["whitespace"], s["weakRivals"]), "",
                "**Target:** %s" % o["target"], "", "**Problem:** %s" % o["problem"], "",
                "**Existing solutions:** %s" % o["existingSolutions"], "", "**Gap:** %s" % o["gap"], "",
                "**Pitch:** %s" % o["pitch"], "", "**MVP scope:**"] + ["- %s" % b for b in o["mvpScope"]] + [
                "", "**First validation step:** %s" % o["firstValidationStep"], ""]
        if o.get("skeptic"):
            out += ["**Skeptic objections:**", ""]
            for j, x in enumerate(o["skeptic"]):
                out += ["%d. %s" % (j + 1, x["objection"]), "   Basis: %s" % x["basis"], "   Would change mind: %s" % x["wouldChangeMind"], ""]
        out += ["---", ""]

    out += ["## Gap status", ""]
    for g in view["gaps"]:
        products = ", ".join(p["name"] for p in g.get("foundProducts", [])) or "—"
        wedge = " Wedge: %s" % g["remainingWedge"] if g.get("remainingWedge") else ""
        out.append("- **%s** — %s (cluster %s). Found: %s%s" % (g["status"], g["unmetNeed"], g["clusterId"], products, wedge))
    out += ["", "## Clusters", ""]
    for c in view["clusters"]:
        out.append("- **%s** — %s. Signals: %s%s" % (c["name"], c["jobToBeDone"], ", ".join(c["signalIds"]), " _(weak)_" if c.get("weak") else ""))
    out += ["", "## Competitors", ""]
    for c in view["competitors"]:
        rating = "%s★" % c["rating"] if c.get("rating") is not None else "n/a"
        complaints = "; complaints: %s" % "; ".join(x["text"] for x in c["complaints"]) if c.get("complaints") else ""
        out.append("- **%s** (%s) — %s, %s%s" % (c["name"], c["category"], c.get("pricing") or "pricing n/a", rating, complaints))
    out += ["", "## Signals", ""]
    for s in view["signals"]:
        out.append('- [%s, intensity %s] %s _(%s)_ "%s"' % (s["type"], s["intensity"], s["statement"], ", ".join(s["evidenceIds"]), s["quote"]))
    out += ["", "## Research trace", ""]
    for c in view["searchCalls"]:
        p = c["params"]
        q = p.get("q") or p.get("search_query") or p.get("product_id") or ""
        kind = c["status"] if c.get("status", "ok") != "ok" else ("cached" if c["cached"] else "live")
        out.append("- `%s` — %s · %d rows · %s · %dms" % (c["engine"], q, c.get("resultCount", 0), kind, c["latencyMs"]))
    out += ["", "## Evidence", ""]
    for e in evidence:
        url = " — %s" % e["url"] if e.get("url") else ""
        snip = ' · "%s"' % e["snippet"] if e.get("snippet") else ""
        out.append("- [%s] (%s · %s) %s%s%s" % (e["id"], e["blockType"], e.get("domain", ""), e.get("title", ""), url, snip))
    return "\n".join(out)

File: src/lib/export.ts
import { regionName } from "./config";
import type { RunView } from "./schemas";

/** Server-side Markdown report exported from a completed run view. */
export function exportMarkdown(view: RunView): string {
  const { run, opportunities, clusters, gaps, competitors, signals, evidence, searchCalls } = view;
  const lines: string[] = [];
  const now = new Date(run.finishedAt ?? run.createdAt).toISOString().slice(0, 10);

  lines.push(`# LaunchRadar — ${run.question}`, "");
  lines.push(`- Region: ${regionName(run.region)}`);
  lines.push(`- Run: ${run.id} · ${now}`);
  lines.push(`- Searches used: ${run.searchesUsed} / budget ${run.budget}`);
  lines.push(`- Confidence elements: ${evidence.length} evidence rows, ${new Set(evidence.map((e) => e.domain).filter(Boolean)).size} domains, ${new Set(evidence.map((e) => e.blockType)).size} source types`);
  lines.push("");

  lines.push("## Opportunities (ranked)", "");
  if (opportunities.length === 0) lines.push("_None — every candidate gap is already served._", "");
  opportunities.forEach((o, i) => {
    lines.push(`### ${i + 1}. ${o.title}`, "");
    lines.push(`Score **${o.score}/100** (${o.confidence}). Pain ${o.subScores.pain}, Momentum ${o.subScores.momentum}, Commercial ${o.subScores.commercial}, Whitespace ${o.subScores.whitespace}, Weak rivals ${o.subScores.weakRivals}.`, "");
    lines.push(`**Target:** ${o.target}`, "");
    lines.push(`**Problem:** ${o.problem}`, "");
    lines.push(`**Existing solutions:** ${o.existingSolutions}`, "");
    lines.push(`**Gap:** ${o.gap}`, "");
    lines.push(`**Pitch:** ${o.pitch}`, "");
    lines.push("**MVP scope:**", ...o.mvpScope.map((b) => `- ${b}`), "");
    lines.push(`**First validation step:** ${o.firstValidationStep}`, "");
    if (o.skeptic.length > 0) {
      lines.push("**Skeptic objections:**", "");
      o.skeptic.forEach((s, j) => {
        lines.push(`${j + 1}. ${s.objection}`);
        lines.push(`   Basis: ${s.basis}`);
        lines.push(`   Would change mind: ${s.wouldChangeMind}`, "");
      });
    }
    lines.push("---", "");
  });

  lines.push("## Gap status", "");
  gaps.forEach((g) => {
    const products = g.foundProducts.map((p) => p.name).join(", ") || "—";
    lines.push(`- **${g.status}** — ${g.unmetNeed} (cluster ${g.clusterId}). Found: ${products}${g.remainingWedge ? ` Wedge: ${g.remainingWedge}` : ""}`);
  });
  lines.push("");

  lines.push("## Clusters", "");
  clusters.forEach((c) => {
    const sigs = c.signalIds.join(", ");
    lines.push(`- **${c.name}** — ${c.jobToBeDone}. Signals: ${sigs}${c.weak ? " _(weak)_" : ""}`);
  });
  lines.push("");

  lines.push("## Competitors", "");
  competitors.forEach((c) => {
    const rating = c.rating !== null ? `${c.rating}★` : "n/a";
    const complaints = c.complaints.length > 0 ? `; complaints: ${c.complaints.map((x) => x.text).join("; ")}` : "";
    lines.push(`- **${c.name}** (${c.category}) — ${c.pricing ?? "pricing n/a"}, ${rating}${complaints}`);
  });
  lines.push("");

  lines.push("## Signals", "");
  signals.forEach((s) => {
    lines.push(
      `- [${s.type}, intensity ${s.intensity}] ${s.statement} _(${s.evidenceIds.join(", ")})_ "${s.quote}"`,
    );
  });
  lines.push("");

  lines.push("## Research trace", "");
  searchCalls.forEach((c) => {
    lines.push(`- \`${c.engine}\` — ${String(c.params.q ?? c.params.search_query ?? c.params.product_id ?? "")} · ${c.resultCount} rows · ${c.cached ? "cached" : "live"} · ${c.latencyMs}ms`);
  });
  lines.push("");

  lines.push("## Evidence", "");
  evidence.forEach((e) => {
    lines.push(`- [${e.id}] (${e.blockType} · ${e.domain}) ${e.title}${e.url ? ` — ${e.url}` : ""}${e.snippet ? ` · "${e.snippet}"` : ""}`);
  });

  return lines.join("\n");
}

export function markdownFileName(runId: string): string {
  return `launchradar-${runId}.md`;
}